### DISCUSSION TITLE
Support ES6

### DISCUSSION THREAD

**Discussion by kalwinskidawid** [apache/cordova#434] [Q&A]
Hi,
Does cordova-android support es6? I mean I know it's just chrome engine, so the engine supports ES6. However, when I type `"type: "module"` in `package.json` then I get the error `Could not load API for android project C:\n[...]\nCordova API.js`. 
Is there any way to be able to use ES6 in cordova? 
For example:
I have `class Firebase { ... }` and I'd like to `import Firebase from './firebase.js'` in my `app.js`

**Comment by breautek** [comment]
There's kind of different environments at play here.

Short answer: No

The webview itself may ES Modules support. Chrome has ESM support since Chrome 61, so as long as the device is running Chrome 61+, you should be able to use `<script type="module=" src="myModule.mjs"></script>`, so your application itself might be able to consume ES modules, but I honestly not even sure if this works (cause I know they have specific mime requirements, which I'm honestly not sure if cordova fulfils). Factory API 24 devices will ship with Chrome 53 (based on the AOSP simulators), so assuming that device is running Chrome 61+ might be an unsafe assumption.

But for everything else there is a lot of ESM interop issues between CommonJS and ESM.

Plugins impementing JS modules for cordova environments uses a module system that predates CommonJS but is very much similar to commonjs (in the sense that they have a `exports` object and `require` method). The difference is the exported object is what is used to clobber the global namespace. So ESM support won't work well here, because cordova JS module can't have an asynchronous export definition. Even if a plugin exports an object that is then modified by reference later in time would simply cause `deviceready` to fire before that plugin is actually ready to be used.

For the Cordova tooling itself that runs in NodeJS, there's a lot of interop issues between CommonJS and ESM. Cordova tooling was built with the module system available at the time, which was commonjs format. Though I can't think of concrete examples off the top of my head, I do know that it was preventing us from upgrading certain dependencies as they moved to ESM.

**Reply by breautek** [reply]
It's unclear where you're trying to use ESM modules but if you're trying to make a cordova plugin with an ESM module, then I think you'll run into a lot of issues that I don't think can be solved without introducing huge breaking changes to cordova's module system and that's something I don't foresee happening, because community tends to favour not breaking existing plugins... a lot of which have been untouched/unmaintained now so introducing breaking changes to those plugins tends to become a huge problem.

**Reply by breautek** [reply]
Also note that... `package.json` module key refers to the node module, but your app code doesn't run in a node runtime, it runs in the browser.

But it will cause cordova tools to be ran in node as if they are ES modules which they are not. Which would be why you see

> However, when I type "type: "module" in package.json then I get the error Could not load API for android project C:\n[...]\nCordova API.js.

**Reply by kalwinskidawid** [reply]
Thanks!
